spinalcordtoolbox/__init__.py

```diff
@@ -1,4 +1,21 @@
 #!/usr/bin/env python
 
-# convenience and backwards compat
+# explicitely expose __version__ and friends in top-level package
 from .utils import __version__, __sct_dir__, __data_dir__, __deepseg_dir__
+
+from . import aggregate_slicewise
+from . import cropping
+from . import download
+from . import image
+from . import labels
+from . import math
+from . import metadata
+from . import flattening
+from . import process_seg
+from . import resampling
+from . import straightening
+from . import template
+from . import types
```

Comment:
Importing a package implicitly runs the `__init__.py` inside the package. If you take a look at the sub-packages they also import their respective relative modules inside their `__init__.py`.

More info here: https://docs.python.org/3/reference/import.html#regular-packages

Reply:
If I'm understanding correctly, though, when we run `import spinalcordtoolbox` then currently, `sct.deepseg`, etc. won't be available by default. Only what we've explicitly named here? 

```python
>>> import spinalcordtoolbox as sct
>>> sct.download
Out[3]: <module 'spinalcordtoolbox.download' from '/home/joshua/Repos/spinalcordtoolbox/spinalcordtoolbox/download.py'>
>>> sct.download.download_data
Out[4]: <function spinalcordtoolbox.download.download_data(urls)>
>>> sct.deepseg
Traceback (most recent call last):
  File "/home/joshua/Repos/spinalcordtoolbox/python/envs/venv_sct/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3343, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-5-deddbddf3fe5>", line 1, in <module>
    sct.deepseg
AttributeError: module 'spinalcordtoolbox' has no attribute 'deepseg'
```

This is what looks to be available with `import spinalcordtoolbox`:

![Screenshot from 2020-10-30 09-32-42](https://user-images.githubusercontent.com/16181459/97711003-e2f22700-1a92-11eb-8a12-06b34158220c.png)

Reply:
```
(venv_sct) ➜  spinalcordtoolbox git:(aj/2963-fix-api-importing-scripts) git checkout master                           
Switched to branch 'master'
Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
(venv_sct) ➜  spinalcordtoolbox git:(master) python
Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import spinalcordtoolbox as sct
>>> dir(sct)
['__builtins__', '__cached__', '__data_dir__', '__deepseg_dir__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__sct_dir__', '__spec__', '__version__', 'utils']
>>> 
(venv_sct) ➜  spinalcordtoolbox git:(master) git checkout aj/2963-fix-api-importing-scripts
Switched to branch 'aj/2963-fix-api-importing-scripts'
(venv_sct) ➜  spinalcordtoolbox git:(aj/2963-fix-api-importing-scripts) python
Python 3.6.12 |Anaconda, Inc.| (default, Sep  8 2020, 23:10:56) 
[GCC 7.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import spinalcordtoolbox as sct
>>> dir(sct)
['__builtins__', '__cached__', '__data_dir__', '__deepseg_dir__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__sct_dir__', '__spec__', '__version__', 'aggregate_slicewise', 'centerline', 'cropping', 'download', 'flattening', 'image', 'labels', 'math', 'metadata', 'process_seg', 'resampling', 'straightening', 'template', 'types', 'utils']
```

Reply:
I don't see how you have access to `sct.download` on `master` without a fully qualified import. Maybe because your environment isn't 100% clean? Maybe you have `spinalcordtoolbox/spinalcordtoolbox/` in your `PATH`? 

You can try in a fresh environment:
```
# from inside sct dir
docker run -v /tmp/spinalcordtoolbox:/spinalcordtoolbox --interactive --tty --entrypoint /bin/bash --net=host python:3.6
# run install_sct and try again?
```

Reply:
I'm not on `master`? What you're getting [here](https://github.com/neuropoly/spinalcordtoolbox/pull/2966#discussion_r515104647) for both branches is what I get too. :)

That's sort of a different thing than what I was asking, heh.

Reply:
We export the subpackages via their own `__init__.py`. Python looks recursively for `__init__.py` files when importing a package. If it sees an `__init__.py` in a sub-folder it runs it.

e.g: https://github.com/neuropoly/spinalcordtoolbox/blob/c8af6cc7bc6898ec33c9670a89cf04cafccdb6a6/spinalcordtoolbox/deepseg/__init__.py

Reply:
But that means we can't import just `spinalcordtoolbox` and access `deepseg`, then? i.e. we have to explicitly import `deepseg` itself?

Reply:
Before you couldn't, now you can (after this).

See relevant: https://github.com/neuropoly/spinalcordtoolbox/pull/2969#discussion_r511924214

^ same concept but a level higher

Reply:
Sorry, I still don't think I have explained myself very well...

> See relevant: https://github.com/neuropoly/spinalcordtoolbox/pull/2969#discussion_r511924214

The usage there still requires you to import `deepseg` explicitly, using `import spinalcordtoolbox.deepseg`.

What I'm talking about is just importing `spinalcordtoolbox` by itself, then having access to the subpackages immediately. i.e. the following:

```python
>>> import spinalcordtoolbox as sct
>>> sct.deepseg
<module 'spinalcordtoolbox.deepseg' from '/home/joshua/Repos/spinalcordtoolbox/spinalcordtoolbox/deepseg/__init__.py'>
>>> sct.deepseg.models.list_models()
{'t2star_sc': {'url': ['https://github.com/ivadomed/t2star_sc/releases/download/r20200622/r20200622_t2star_sc.zip',
   'https://osf.io/v9hs8/download?version=5'],
  'description': 'Cord segmentation model on T2*-weighted contrast.',
  'default': True},
```

So, I am suggesting to possibly add `from . import deepseg` (along with the other subpackages) to the top-level package `__init__.py`.

Reply:
The `import spinalcordtoolbox` having access to submodules immediately should be subject to discussion in another issue.

Reply:
@zougloub Isn't giving access to submodules immediately the whole purpose of this PR? From @Drulex's https://github.com/neuropoly/spinalcordtoolbox/pull/2966#discussion_r511195929

> Yes and same goes for any API module once we merge [f65df32](https://github.com/neuropoly/spinalcordtoolbox/commit/f65df321b078b939e5de66755ca0561fe773ed82) because it allows more sane imports like `import spinalcordtoolbox as sct` with access to all sub-packages without explicit importing. For example you expect to do:
> 
> ```
> import os
> os.path.join(...)
> ```
> 
> instead of:
> 
> ```
> import os
> import os.path
> import os.path.join
> os.path.join(...)
> ```
> 
> Right now on `master` we can only do the second form, this PR combined with #2955 allows the first form which is standard practice.

Reply:
Sorry @joshuacwnewton I misunderstood what you were saying. Indeed this PR does not propose to implicitly expose sub-packages, but only modules of the `spinalcordtoolbox` package. In the PR's current form sub-packages would still need to be imported explicitly. I created an issue to discuss since it seems to be a hot topic: #2989